### PR TITLE
Data: move element cache to element[expando] (updated)

### PR DESCRIPTION
**Pull request by rwaldron**
- use Object.defineProperty
- remove unnecessary code path

Signed-off-by: Rick Waldron <waldron.rick@gmail.com>

### CODE REVIEW

**Comment by rwaldron** [comment]
cc @gibson042

File: src/data/Data.js [hunk 1]
```diff
@@ -27,12 +27,12 @@ Data.prototype = {
...
 			// deleted with the delete operator
 			} else {
-				descriptor[ this.expando ] = {
+				descriptor = {
 					value: value,
 					writable: true,
 					configurable: true
 				};
-				Object.defineProperties( owner, descriptor );
+				Object.defineProperties( owner, this.expando, descriptor );
```
Comment (rwaldron) [inline src/data/Data.js]: Derp.

File: src/data/Data.js [hunk 2]
```diff
@@ -27,12 +27,12 @@ Data.prototype = {
...
 			// deleted with the delete operator
 			} else {
-				descriptor[ this.expando ] = {
+				descriptor = {
 					value: value,
 					writable: true,
 					configurable: true
 				};
-				Object.defineProperties( owner, descriptor );
+				Object.defineProperty( owner, this.expando, descriptor );
```
Comment (gibson042) [inline src/data/Data.js]: This whole block can be simpler, and doesn't even need a `try`; see 95fb798980d7e404c413e29e20016db9052e2bf2.

File: src/data/Data.js [hunk 3]
```diff
@@ -128,7 +122,7 @@ Data.prototype = {
 	},
 	remove: function( owner, key ) {
 		var i, name, camel,
-			cache = this.cache( owner );
+			cache = owner[ this.expando ];
```
Comment (gibson042) [inline src/data/Data.js]: We need an early return when `cache` is nonexistent to avoid errors, right?

File: src/data/Data.js (hunk 2 above)
Comment (jbedard) [inline src/data/Data.js]: And do we want the `descriptor` variable when it is only assigned and used once? Currently it is also assigned twice but only used in this one spot...

Comment (rwaldron) [inline src/data/Data.js]: > This whole block can be simpler, and doesn't even need a try; see 95fb798.

Simplified.

> And do we want the descriptor variable when it is only assigned and used once

Allocation removed.

File: src/data/Data.js (hunk 3 above)
Comment (rwaldron) [inline src/data/Data.js]: I didn't observe any such failures, but an early return is a nice optimization that doesn't appear to have any adverse side effects

File: src/data/Data.js [hunk 4]
```diff
@@ -14,34 +14,25 @@ Data.accepts = jQuery.acceptData;
...
-		// Support: Android < 4
-		// Fallback to a less secure definition
-		} catch ( e ) {
-			descriptor[ this.expando ] = value;
-			jQuery.extend( owner, descriptor );
+		// Otherwise secure it in a non-enumerable, non-writable property
+		// configurability must be true to allow the property to be
+		// deleted with the delete operator
+		} else {
+			descriptor = {
```
Comment (jbedard) [inline src/data/Data.js]: Is there a reason to keep the variable? Why not just
```javascript
Object.defineProperty( owner, this.expando, {
	value: value,
	writable: true,
	configurable: true
});
```
(which I see in a few other places in the code, not sure what the convention is though...)

Comment (rwaldron) [inline src/data/Data.js]: Left over from when the same object was used in two places, but yes: I've just now removed that and changed to your suggestion.

**Comment by gibson042** [comment]
LGTM :+1:

**Comment by rwaldron** [comment]
@gibson042 Cool, thanks for the feedback/patience/attention ;)

**Pull request closed without merging**